attempt using:

http://scikit-learn.org/stable/auto_examples/neural_networks/plot_mnist_filters.html#sphx-glr-auto-examples-neural-networks-plot-mnist-filters-py

In [2]:
# import libraries
import numpy as np
import csv
from sklearn.neural_network import MLPClassifier
from sklearn.model_selection import GridSearchCV

In [3]:
# Read in the data
# Create the training and test sets, skipping the header
# row with [1:]
dataset = np.genfromtxt(open('train.csv', 'r'),
                        delimiter=',', dtype='f8')[1:]
target = [x[0] for x in dataset]
train = [x[1:]/255 for x in dataset]
test = np.genfromtxt(open('test.csv', 'r'),
                     delimiter =',', dtype='f8')[1:]
test = [x/255 for x in test]

In [4]:
# set model
mlp = MLPClassifier(
    early_stopping=True, 
    hidden_layer_sizes=(50,50)
)

In [5]:
# hyperparameters
param_grid = {
    'hidden_layer_sizes': [
        (2000), (1200, 800, 400, 200)
    ],
    'max_iter': [200, 400],
    'learning_rate': ['constant', 'adaptive'] 
}

clf = GridSearchCV(mlp, param_grid)
clf.fit(train, target)
print('Best score: {}'.format(clf.best_score_))
print('Best parameters: {}'.format(clf.best_params_))

Best score: 0.975928571429
Best parameters: {'learning_rate': 'adaptive', 'max_iter': 200, 'hidden_layer_sizes': 2000}


Previous result:
- Best score: 0.973357142857
- Best parameters: {'hidden_layer_sizes': 800}

- Best score: 0.974071428571
- Best parameters: {'hidden_layer_sizes': 2000}

- Best score: 0.975523809524
- Best parameters: {'learning_rate': 'constant', 'max_iter': 200, 'hidden_layer_sizes': 2000}


- Best score: 0.975928571429
- Best parameters: {'learning_rate': 'adaptive', 'max_iter': 200, 'hidden_layer_sizes': 2000}

In [6]:
clf.score(train, target)

0.99811904761904757

In [10]:
# fit model
#mlp.fit(train, target)
#print(mlp.score(train, target))

In [9]:
# predictions for test set
predictions = clf.predict(test)

In [10]:
# save predictions to file
output = csv.writer(open('results-v3.csv', 'wt'))
output.writerow(["ImageId", "Label"])
for i, prediction in enumerate(predictions):
    output.writerow([i+1, int(prediction)])